### PR TITLE
Added dag functions and fixed PEP8 errors

### PR DESCRIPTION
**Pull request by Erotemic**
Added transitive_reduction, source_nodes, and sink_nodes.
Added corresponding tests.
Fixed minor PEP8 Errors.

In my projects I've written a bunch of helper functions for networkx and I'd like to start merging 
some of the more complete functions into the master branch. 

The main part of this pull request is an implementation of transitive_reductions for DAGs. 
I also added in helpers for source and sink nodes. I added the corresponding tests and they all seem to pass. I also cleaned up minor PEP8 formatting issues inconsistent with the rest of the file.

### CODE REVIEW

**Comment by hagberg** [comment]
Thanks for suggesting this.  

I usually argue against including simple one-line functions.  It's not that they aren't useful it's that it makes one more function to keep track of (document, test, etc) for not much gain.   A middle ground is to add some examples that show how to write the one-liners.

For the transitive reduction function I'd like to see a pure python implementation (not using scipy).  That might be a lot more efficient.  Here are some ideas
http://stackoverflow.com/questions/17078696/im-trying-to-perform-the-transitive-reduction-of-directed-graph-in-python
http://www.boost.org/doc/libs/1_55_0/libs/graph/doc/transitive_closure.html

**Comment by Erotemic** [comment]
Removing the dependency on scipy is a good idea. I changed then implementation of transitive reduction following the example given in the stack overflow post.

I can also understand that one-line functions do create a lot of documentation and test overhead. However, I would argue that the source and sink functions for DAGs are fundamental enough to justify their inclusion. I think finding the root of a tree is a common enough use case to justify at least the source_nodes function by itself. It is true that topological_sort could be used for this, but again I think that source_nodes is much more clear from a readability standpoint. 

If that argument isn't convincing I'll remove source_nodes, sink_nodes, and their tests from the pull request.

File: networkx/algorithms/dag.py [hunk 1]
```diff
@@ -355,6 +387,108 @@ def transitive_closure(G):
...
+            for gchild in list(TR_scc.successors(child)):
+                # Remove all edges from parent to non-child descendants
+                for descendant in list(nx.dfs_preorder_nodes(TR_scc, gchild)):
+                    if TR_scc.has_edge(parent, descendant):
+                        TR_scc.remove_edge(parent, descendant)
+
+    if has_cycles:
+        # Uncondense graph.
+        mapping = G_scc.graph['mapping']
+        inv_mapping = invert_map(mapping)
```
Comment (jfinkels) [inline networkx/algorithms/dag.py]: Might you be able to use the `networkx.utils.groups()` function that I introduced recently?

File: networkx/algorithms/dag.py [hunk 2]
```diff
@@ -355,6 +387,108 @@ def transitive_closure(G):
...
+    has_cycles = not nx.is_directed_acyclic_graph(G)
+    if has_cycles:
+        # To handle cycles convert to graph of SCCs, find transitive reduction
+        # and then make a directed cycle for every SCC.
+        G_scc = nx.condensation(G)
+    else:
+        G_scc = G
+
+    # Adapted from
+    # http://stackoverflow.com/questions/17078696/transitive-reduction-python
```
Comment (jfinkels) [inline networkx/algorithms/dag.py]: StackOverflow content is distributed under a Creative Commons Attribution-ShareAlike 3.0 license.

**Comment by jfinkels** [comment]
I agree with @hagberg on the one-liners. `sources = [v for v, d for G.in_degree() if d == 0]`.

**Comment by Erotemic** [comment]
It seems that this feature was merged in #2215

**Comment by jfinkels** [comment]
Oops, sorry I didn't make the connection. Thanks for checking up, and please continue to contribute!

**Pull request closed without merging**